### PR TITLE
feat: run warmup on Runtimes and Executor

### LINKED ISSUE #5467: feat: run warmup requests on runtime startup to ensure that service is ready to accept connections
**Describe the feature**

Warm up requests are executed before the service is reported ready so that new incoming requests can be readily served without having to create new network connections to database dependency or sidecar container, etc or load the required modules. Warm up requests can be dummy or canary or health check type requests that can trigger the hot path at least once. Warm up requests don't need to be successful but need to be executed before reporting readiness.

**Your proposal**

Gateway Runtime: execute service discovery request to all executors.
Executor Runtime: execute empty post request on the default endpoint.

---

**Environment**

**Screenshots**

### PR DESCRIPTION
**Pull request by girishc13**
**Goals:**

- resolves #5467 
- Gateway and Head runtime now starts a warmup tasks on the `run_forever` runtime method.
- The warmup task is created by the `GatewayStreamer` or the Head `RequestHandler` that delegates the warmup per deployments to the connection pool.
  - A single task per deployment is created and await until completion.
- The `GrpcConnectionPool` fetches the available shards and or replica (target) channels and creates a `JinaInfoRPC` request task. 
  - One task is created for each channel.
  - The warmup cycle runs until:
    - all targets return a successful response or
    - a max timeout of 5 minutes is reached or
    - a stop event is set by the runtime due to early tear down of the runtime.
  - In the warmup cycle, only unsuccessful targets are attempted.
- `CompositeGateway` doesn't support the warmup due to the creation of separate `GatewayStreamer` objects for each type of Gateway. The issue will be addressed in #5592.
  - the CompositeGateway methods `shutdown`, `run_server`, `setup_server` now handle each gateway operation as a task per created gateway.

### CODE REVIEW

File: jina/serve/networking.py [hunk 1]
```diff
@@ -1114,6 +1113,71 @@ async def task_wrapper():
...
+                    return
+
+                await asyncio.sleep(0.2)
+        except Exception as ex:
+            self._logger.error(f'error with warmup up task: {ex}')
+            return
+
+    def __extract_target_to_channel(self, deployment):
+        replica_set = set()
+        replica_set.update(self._connections.get_replicas_all_shards(deployment))
```
Comment (girishc13) [inline jina/serve/networking.py:1171]: I can extract out the warmup logic into a `Warmer` class but the `self._connections` is the inner class `_ConnectionPoolMap` of the `GrpcConnectionPool`. Does the inner class still make sense?

File: jina/serve/runtimes/gateway/request_handling.py [hunk 2]
```diff
@@ -147,7 +147,7 @@ async def _process_results_at_end_gateway(
                         and not self._gathering_endpoints
                     ):
                         self._gathering_endpoints = True
-                        asyncio.create_task(gather_endpoints(request_graph))
+                        await asyncio.create_task(gather_endpoints(request_graph))
```
Comment (JoanFM) [inline jina/serve/runtimes/gateway/request_handling.py:150]: no please, this is created in the backround and not awaited in purpose

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/serve/networking.py [hunk 3]
```diff
@@ -1,6 +1,8 @@
 import asyncio
 import ipaddress
 import os
+import threading
+import time
```
Comment (JoanFM) [inline jina/serve/networking.py:5]: please remove unneeded imports

File: jina/serve/runtimes/asyncio.py [hunk 4]
```diff
@@ -1,6 +1,7 @@
 import argparse
 import asyncio
 import signal
+import threading
```
Comment (JoanFM) [inline jina/serve/runtimes/asyncio.py:4]: remove unneeded import

File: jina/serve/runtimes/head/request_handling.py [hunk 5]
```diff
@@ -1,6 +1,9 @@
 import asyncio
-from typing import TYPE_CHECKING, Dict, Optional, Tuple
+import threading
```
Comment (JoanFM) [inline jina/serve/runtimes/head/request_handling.py:2]: please cleanup imports

File: jina/serve/networking.py [hunk 6]
```diff
@@ -178,6 +183,10 @@ def add_connection(self, address: str, deployment_name: str):
             stubs, channel = self._create_connection(address, deployment_name)
             self._address_to_channel[address] = channel
             self._connections.append(stubs)
+            # create a new set of stubs and channels for warmup to avoid
+            # loosing channel during remove_connection or reset_connection
+            stubs, _ = self._create_connection(address, deployment_name)
+            self._warmup_stubs.add(stubs)
```
Comment (JoanFM) [inline jina/serve/networking.py:189]: I do not see where they are used

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/serve/networking.py (hunk 6 above)
Comment (girishc13) [reply jina/serve/networking.py:189]: Check the usage of the property method `self.warmup_stubs`. What's the preference when exposing `read only` properties?

Comment (JoanFM) [reply jina/serve/networking.py:190]: ah okey

**Review by JoanFM — APPROVED** [review]

**Pull request merged**